src/ViewModeResolver.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Drupal\format_strawberryfield;
+
+use Drupal\Core\Cache\Cache;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Cache\UseCacheBackendTrait;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\strawberryfield\StrawberryfieldUtilityServiceInterface;
+
+/**
+ * View Mode resolver service.
+ *
+ * @ingroup format_strawberryfield
+ */
+class ViewModeResolver implements ViewModeResolverInterface {
+  use UseCacheBackendTrait;
+
+  /**
+   * The config factory service.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface $configFactory
+   */
+  protected $configFactory;
+
+  /**
+   * The Strawberry Field Utility Service
+   *
+   * @var \Drupal\strawberryfield\StrawberryfieldUtilityService
+   */
+  protected $strawberryfieldUtility;
+
+  /**
+   * DisplayResolver constructor.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   * @param \Drupal\strawberryfield\StrawberryfieldUtilityServiceInterface $strawberryfield_utility_service
+   *   The SBF utility Service
+   */
+  public function __construct(StrawberryfieldUtilityServiceInterface $strawberryfield_utility_service, ConfigFactoryInterface $config_factory) {
+    $this->strawberryfieldUtility = $strawberryfield_utility_service;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCandidates(ContentEntityInterface $entity) {
+    if ($ado_types = $this->getADOTypes($entity)) {
+      $view_modes = array_filter($this->getSortedMappings(), function($mapping) use ($ado_types) {
+        return ($mapping['active'] == TRUE) && in_array($mapping['jsontype'], $ado_types);
+      });
+      return $view_modes;
+    }
+
+    return ['default'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function get(ContentEntityInterface $entity) {
+    $ado_types = $this->getADOTypes($entity);
+    if (!empty($ado_types)) {
+      foreach ($this->getSortedMappings() as $mapping) {
+        if (($mapping['active'] == TRUE) && in_array($mapping['jsontype'], $ado_types)) {
+          return $mapping['view_mode'];
+        }
+      }
+    }
+
+    return 'default';
+  }
+
+  /**
+   * Get a list of ADO types based on the SBF.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity.
+   *
+   * @return array
+   *  Array of ado types.
+   *
+   * @throws \Exception
+   */
+  protected function getADOTypes(ContentEntityInterface $entity) {
+    $cache_id = 'format_strawberry:view_mode_adotypes:' . $entity->id();
```

Comment:
This cache should be per "JSON type" not per entity ID: on a 400,000 entity site the cache instead of helping will become a burden. If ADO or not is at the end a bundle type definition, and which view mode a JSON key "type" value one.

Reply:
@DiegoPino I don't see how this can be cached this way, as we're dependent on the entity. On this list we're caching the ADO types per entity as `bearsStrawberryfield` service still gets it as parameter. Seems like a catch 22, I can't cache before calling and I don't know the necessary information before caching.

Reply:
The egg/chicken dilemma (catch 22) is the reason we did not apply a cache previously for the json type itself when just having an entity. (the lack of a proper class of course at bundle level). Let me think about this a little deeper (like 24 hours, busy day preparing for tomorrow many workshop/zoom type of things) and give the whole thing a spin. I'm pretty sure there may be a pattern that is escaping us where performance and atomicity can play well. Thanks

Reply:
I've been thinking about this for a bit, as if the solution of caching it per entity is not acceptable, having a computed field wouldn't be either, I think that giving how aggressively Drupal caches everything, this would be fine but you obviously know the system better. Drupal performs so so with a lot of content without external caching and for memcache/redis, 400k simple entries like these are a breeze.

In any case, this is providing an additional cache layer that we don't currently have, so I would suggest three things here:

- Removing the cache layer.
- Adding a followup issue where we consider the field approach instead (this would replace the service)
- Having a place where we can actually test this; maybe an online version of a mid sized repository with something like new relic or the symfony one, where we can benchmark stuff?

Reply:
> - Removing the cache layer.
Ok. I agree. Let's remove the cache layer for now. Caching which bundles are of ADO type may something easier we can approach (at least one check less) and that can go into the actual `::bearsStrawberryfield` method (another issue, not to be solved here)
> - Adding a followup issue where we consider the field approach instead (this would replace the service)

The field approach you mean for saving the view mode at the Node level? Instead of using the ds-> config? Yes. may even an option that people want to opt/in/out. We may keep the service and build a field check there if we go that route. Better to have more options than less.

> - Having a place where we can actually test this; maybe an online version of a mid sized repository with something like new relic or the symfony one, where we can benchmark stuff?
I agree. We can feed one of our repos a lot of data and even create a test batch ingest set (via the JSON API, a special costume module) so we can benchmark this. I once created 1,000,000 random objects in 2018 for this (when I decided on this architecture) but it was a brute force approach. Should be see this as a DevOps task with @alliomeria collaboration for the data?

Reply:
I will mark the cache discussion as resolved for now until we have larger numbers and we can measure if cache v/s dynamically calculating this every time changes or not things. Thanks a lot of https://github.com/esmero/format_strawberryfield/pull/122/commits/bc84fe57d9fea3ca18f73f719159b824254bfc4d Which was needed.